glib/src/gstring.rs

```diff
@@ -14,6 +14,326 @@ use std::ptr;
+    }
+    // rustdoc-stripper-ignore-next
+    /// Wraps a raw C string with a safe GLib string wrapper. The provided C string **must** be
+    /// valid UTF-8 and nul-terminated. All constraints from [`std::ffi::CStr::from_ptr`] also
+    /// apply here.
+    #[inline]
+    pub unsafe fn from_ptr<'a>(ptr: *const c_char) -> &'a Self {
+        let cstr = CStr::from_ptr(ptr);
+        Self::from_bytes_with_nul_unchecked(cstr.to_bytes_with_nul())
+    }
+    // rustdoc-stripper-ignore-next
+    /// Converts this GLib string to a byte slice containing the trailing 0 byte.
+    ///
+    /// This function is the equivalent of [`GStr::to_bytes`] except that it will retain the
+    /// trailing nul terminator instead of chopping it off.
+    #[inline]
+    pub fn to_bytes_with_nul(&self) -> &[u8] {
+        &self.0
+    }
+    // rustdoc-stripper-ignore-next
+    /// Converts this GLib string to a byte slice.
+    ///
+    /// The returned slice will **not** contain the trailing nul terminator that this GLib
+    /// string has.
+    #[inline]
+    pub fn to_bytes(&self) -> &[u8] {
+        &self.0[0..self.0.len() - 1]
+    }
+    // rustdoc-stripper-ignore-next
+    /// Returns the inner pointer to this GLib string.
+    ///
+    /// The returned pointer will be valid for as long as `self` is, and points to a contiguous
+    /// region of memory terminated with a 0 byte to represent the end of the string.
+    ///
+    /// **WARNING**
+    ///
+    /// The returned pointer is read-only; writing to it (including passing it to C code that
+    /// writes to it) causes undefined behavior. It is your responsibility to make
+    /// sure that the underlying memory is not freed too early.
+    #[inline]
+    pub fn as_ptr(&self) -> *const c_char {
+        self.0.as_ptr() as *const _
+    }
+    // rustdoc-stripper-ignore-next
+    /// Converts this GLib string to a string slice.
+    #[inline]
+    pub fn as_str(&self) -> &str {
+        unsafe { std::str::from_utf8_unchecked(self.to_bytes()) }
+    }
+    // rustdoc-stripper-ignore-next
+    /// Converts this GLib string to a C string slice.
+    #[inline]
+    pub fn as_c_str(&self) -> &CStr {
+        unsafe { CStr::from_bytes_with_nul_unchecked(&self.0) }
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// Converts a static string literal into a static nul-terminated string.
+///
+/// The expanded expression has type [`&'static GStr`]. This macro will panic if the
+/// string literal contains any interior nul bytes.
+///
+/// # Examples
+///
+/// ```
+/// # fn main() {
+/// use glib::{gstr, GStr, GString};
+///
+/// const MY_STRING: &GStr = gstr!("Hello");
+/// assert_eq!(MY_STRING.to_bytes_with_nul()[5], 0u8);
+/// let owned: GString = MY_STRING.to_owned();
+/// assert_eq!(MY_STRING, owned);
+/// # }
+/// ```
+///
+/// [`&'static GStr`]: crate::GStr
+#[macro_export]
+macro_rules! gstr {
+    ($s:literal) => {
+        unsafe { $crate::GStr::from_bytes_with_nul_unchecked($crate::cstr_bytes!($s)) }
```

Comment:
Why so complicated?

```suggestion
        unsafe { $crate::GStr::from_bytes_with_nul_unchecked(concat!($s, "\0").as_bytes()) }
```

should also work or not?

Reply:
We need to check at compile time if the string literal has any interior nul bytes, because otherwise it breaks the invariant on `GString` where it has to be a valid `CString` internally. The only way to do that is a proc macro. The rest I put in a `macro_rules` to get the hygiene with `$crate`.

Reply:
Ah I see